Gradient view and more colormaps

squirrelo opened the pull request:
~~Requires/supersedes #455 depending on how we want to do reviews.~~ Merged

This updates the colorview controller to allow gradient-based coloring instead of single-step coloring. It also adds in every available colormap in chroma into emperor.

This also fixes an issue found where this.$body was pointing to the slickgrid div itself, instead of a container div for the body with the slickgrid in it. Fixing this allowed the switch between colorbar and discrete category colors in the menu.

<img width="485" alt="screen shot 2016-04-15 at 19 18 27" src="https://cloud.githubusercontent.com/assets/3079066/14578534/5482ebcc-033f-11e6-953b-0e618b427b6a.png">

<img width="1664" alt="screen shot 2016-04-15 at 19 18 47" src="https://cloud.githubusercontent.com/assets/3079066/14578535/59a9abfe-033f-11e6-8d1c-a7dd17c4b72f.png">

ElDeveloper commented on emperor/support_files/js/color-view-controller.js:
```diff
@@ -56,62 +55,108 @@ define([
     var name, value, colorItem;
 
+    // Create scale div and checkbox for whether using scalar data or not
+    this.$scaleDiv = $('<div></div>')
```
Can you describe these properties in the documentation above? Also maybe some of these need to be private properties, I'll leave that to you, let me know if you would like to talk through that.

squirrelo replied:
Done

ElDeveloper commented on emperor/support_files/js/color-view-controller.js:
```diff
@@ -56,62 +55,108 @@ define([
+    this.$scaleDiv = $('<div></div>')
+    this.$colorScale = $("<svg width='90%' height='40'></svg>");
+    this.$scaleDiv.append(this.$colorScale);
+    this.$scaled = $("<input type='checkbox' id='scaled'>");
```
Unfortunately we cannot use the id tag, because in a Jupyter notebook we may have `n` different plots, so this would conflict between plots.

squirrelo replied:
Yeah, I gave it an id but don't use it anywhere. I can remove that.

wasade replied:
Would it make sense to have a plot-specific prefix that's a randomly generated tag provided to all IDs? Not sure if this would work or if the syntax is sane but:

```javascript
function id_singleton_prefix() { 
    if(global_prefix === null) {
        global_prefix = something_random;
    return global_prefix;
}

...
this.$scaled = $("<input type='checkbox' id=id_singleton_prefix() + 'scaled'>");
```

squirrelo replied:
As far as I can tell in the codebase (and @ElDeveloper correct me if I'm wrong) nothing is done through IDs directly, but instead using the objects attached to the controllers, so this is probably not needed.

ElDeveloper replied:
Yeah, you are right Josh, for these properties, there's no need to have
unique identifiers. However, we have these for the
EmperorViewControllers in general, we have an `identifier` property and
it follows a scheme similar to what Daniel suggests.

On (Apr-18-16| 9:59), Joshua Shorenstein wrote:
> > @@ -56,62 +55,108 @@ define([
> >      var SLICK_WIDTH = 25, scope = this;
> >      var name, value, colorItem;
> >  
> > +    // Create scale div and checkbox for whether using scalar data or not
> > +    this.$scaleDiv = $('<div></div>')
> > +    this.$colorScale = $("<svg width='90%' height='40'></svg>");
> > +    this.$scaleDiv.append(this.$colorScale);
> > +    this.$scaled = $("<input type='checkbox' id='scaled'>");
> 
> As far as I can tell in the codebase (and @ElDeveloper correct me if I'm wrong) nothing is done through IDs directly, but instead using the objects attached to the controllers, so this is probably not needed.
> 
> ---
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/emperor/pull/456/files/c8896fbb1b35090514974c949912f38de068eb3b#r60094408

ElDeveloper commented on emperor/support_files/js/color-view-controller.js:
```diff
@@ -56,62 +55,108 @@ define([
+    for (var i = 0; i < ColorViewController.Colormaps.length; i++) {
+      var colormap = ColorViewController.Colormaps[i];
+      //Check if we are in a new optgroup
+      if (colormap.type != currType) {
```
Can you use `!==`?

squirrelo replied:
yup

ElDeveloper commented on emperor/support_files/js/color-view-controller.js:
```diff
@@ -128,48 +173,79 @@ define([
+   * @param {discrete} Whether to treat colormap requested as a discrete set
+   * of colors or use interpolation to create gradient of colors
+   * @param {scaled} Whether to use a scaled colormap or equidistant colors for
+   * each value. Defaiult false.
```
Defaiult -> Default is

squirrelo replied:
fixed

ElDeveloper commented on emperor/support_files/js/color-view-controller.js:
```diff
@@ -128,48 +173,79 @@ define([
     }
 
-    return colors;
+    //if scaled, generate the scale and add to div.
```
can you add a space before the `if`?

squirrelo replied:
done

ElDeveloper commented on emperor/support_files/js/color-view-controller.js:
```diff
@@ -128,48 +173,79 @@ define([
+      var mid = (min + max) / 2;
+      step = (max - min) / 100;
+      var stopColors = [];
+      for (var s = min; s <= max; s += step) {
```
I believe you can replace some of this with underscore's range, see here:
http://underscorejs.org/#range

squirrelo replied:
I could do 
```javascript
var range = _.range(min, max+1, step);
for(var pos in range) {
stopColors.push(interpolator(range[pos]).hex());
}
```
but this seems to be more complicated than the current setup.

ElDeveloper commented on emperor/support_files/js/view-controller.js:
```diff
@@ -83,7 +83,7 @@ define([
     this.$header.css('margin', '0 auto');
     this.$header.css('width', '100%');
 
-    this.$body = $('<div name="emperor-view-controller-body"></div>');
```
I believe this should still be `emperor-view-controller-body`, the gridDiv in turn should have a different name i.e. `emperor-grid-div` or something like that.

squirrelo replied:
good point, changed.

ElDeveloper commented:
@squirrelo **THIS IS AMAZING**.

Just a few notes: there's an error in the JavaScript console, though I haven't been able to trigger this again, do you have any idea as to why this may be happening:

```javascript
color-view-controller.js:204 Uncaught TypeError: Cannot read property '0' of undefined
```

Some notes, the `Continuous values` checkbox should be hidden if the color set is in the `Discrete` group, once you switch over to one of `Converging` or `Diverging` it should then be shown.

I belive the `Converging` group should be renamed as `Sequential`.

The `Continuous values` label should read something different, or have an explanation of what it does, maybe we can ask others in the lab what would they call this.

Does it make sense to have the colormap show regardless of what your scheme is i.e. whether the checkbox is marked or not?

The colorbar has a bit of unused space to the right, I think it would make sense to expand that to fit the width.

Finally, would you mind adding tests for the new properties that you added?

squirrelo commented:
Right now the continuous values checkbox is disabled if you are in a discrete colormap, but if really wanted I can make it disappear completely.

For me, it doesn't make sense to show the colormap regardless, because the entire coloring of the data is changed depending on if the slickgrid or colormap is shown. It would get confusing, since if the slickgrid is shown the colors of the points would not match up with the colormap.

I left the space on the right because as you shrink the page, that space gets eaten up. This was the largest I could get the bar and still have it show completely when you shrink the page as small as it goes.

I've added/updated tests for the new and changed properties, but if any are missing I can add them.

wasade commented on emperor/support_files/js/color-view-controller.js:
```diff
@@ -56,62 +55,108 @@ define([
+            scope.$colorScale.html(colorInfo[1]);
+          } else {
+            scope.setSlickGridDataset(data);
+            scope.$scaleDiv.hide();
```
this and the next two lines appear replicated in the if block, do these need to be branch specific (e.g., due to execution order)?

squirrelo replied:
They are reversed, so this is hiding the scale and showing the slickgrid, while the if above does the opposite.

ElDeveloper commented:
> Right now the continuous values checkbox is disabled if you are in a discrete colormap, but if really wanted I can make it disappear completely.

Making it disappear would be better, the checkbox serves no purpose except when the colormap is appropriate.

> For me, it doesn't make sense to show the colormap regardless, because the entire coloring of the data is changed depending on if the slickgrid or colormap is shown. It would get confusing, since if the slickgrid is shown the colors of the points would not match up with the colormap.

Yes I understand what you are saying. Anyone else have any comments on this?

> I left the space on the right because as you shrink the page, that space gets eaten up. This was the largest I could get the bar and still have it show completely when you shrink the page as small as it goes.

Noted, perhaps the only thing needed is to center the colormap, in the space it has?

-----------

Thanks for making these changes!

wasade commented on emperor/support_files/js/color-view-controller.js:
```diff
@@ -128,48 +173,79 @@ define([
    * category in a given metadata column.
    * @param {map} name of the color map to use, see
    * ColorViewController.Colormaps
+   * @param {discrete} Whether to treat colormap requested as a discrete set
```
is there a default?

squirrelo replied:
Nope, since it is a crucial piece of info depending on the colormap used.

wasade commented on emperor/support_files/js/color-view-controller.js:
```diff
@@ -128,48 +173,79 @@ define([
+        max = _.max(numericValues);
+        interpolator = chroma.scale(map).domain([min, max]);
+      }
     }
```
I'm not seeing where `interpolator` get's set if `discrete === true`. From my read of the if/elifs starting on line 215, you _might_ expect the function call to be attempted in that case but I'm also not sure how JS implements if/elif and whether multiple branches can be taken.

squirrelo replied:
There is no interpolator for discrete values, since no interpolation is needed. Like 215 is indeed where the check is made, and javascript respects if/elif and only does one branch. Switch is the only thing that can fall through as far as I'm aware.

wasade commented on emperor/support_files/js/color-view-controller.js:
```diff
@@ -221,34 +297,53 @@ define([
-  "#b2df8a", "#33a02c", "#fb9a99", "#e31a1c", "#fdbf6f", "#ff7f00", "#cab2d6",
-  "#6a3d9a", "#ffff99", "#b15928"];
+  ColorViewController.Colormaps = [
+    {id: 'discrete-coloring-qiime', name: 'Classic QIIME Colors', type: 'Discrete'},
```
minor, but would it be possible to place the `"Discrete"`, `"Converging"`, and `"Diverging"` strings into single variables? Reduces the potential for a copypasta error now or in the future in the event these strings change

squirrelo replied:
I've moved them to headers of each section as objects with `type` property. However, this means there are two different types of objects in this list, so that will need to be taken into account my anything else using this list.

squirrelo commented:
Comments addressed

ElDeveloper commented:
👏
